addons/api/addon/serializers/application.js

```diff
@@ -74,18 +75,33 @@ export default class ApplicationSerializer extends RESTSerializer {
       delete json[key];
     }
     // If an attribute has a `for` option, it must match the
-    // record's `type`, else the attribute is excluded
+    // record's `type`, otherwise the attribute is excluded
     // from serialization.
-    //note: This doesn't handle secrets yet, we can add support for them if needed.
-    if (
-      options?.for &&
-      options.for !== recordType &&
-      !options.for.includes(recordType)
-    ) {
-      if (options.isNestedAttribute) {
-        delete json?.attributes?.[key];
-      } else {
-        delete json[key];
+
+    if (options?.for) {
+      // Check the typeof the `for` option to determine if it is a plugin or not
+      const isPlugin =
+        typeOf(options.for) === 'object' && options.for?.type === 'plugin';
+      const isNotPlugin = typeOf(options.for) === 'string';
+
+      // Helper function to handle deleting attribute and secret keys
+      const deleteKey = (json) => {
+        if (options.isNestedAttribute) {
+          delete json?.attributes?.[key];
+        } else if (options.isNestedSecret && json?.secrets?.[key]) {
+          delete json?.secrets?.[key];
+        } else {
+          delete json[key];
+        }
+      };
+
+      // For plugins, the `for` option is an object with a `name` key
+      // For non-plugins, the `for` option is a string
+      if (
+        (isPlugin && !options.for?.name.includes(compositeType)) ||
+        (isNotPlugin && !options.for?.includes(recordType))
```

Comment:
This check still works but `for` can also be an array for non plugins, we should probably note that in the comment still.

The only edge case could be if `recordType` was a substring of `for` and we would have a bug but I'm not sure if we'd hit that. Maybe it's still worth being explicit and checking for string and `===` vs array and `includes`?

Reply:
yeah you are right, I will include the array for this check and update the comment as well